### ISSUE TITLE
Console Menubar incorrectly added to system menubar

### ISSUE THREAD

**Issue by latenitefilms** [fcpxhacks/fcpxhacks#789]
See:

![screen shot 2017-10-03 at 6 06 15 pm](https://user-images.githubusercontent.com/22286696/31113558-9d013b8e-a865-11e7-9422-da8a45a5012d.png)

**Comment by latenitefilms** [comment]
Not sure how this is happening... Searching all our code for `menubar.new`, doesn't reveal anything obvious?

**Comment by latenitefilms** [comment]
Maybe we could just explicitly tell it `:removeFromMenuBar()`?

**Comment by randomeizer** [comment]
Hmm, I'm having trouble reproducing this one. How do you get it to appear?

**Comment by latenitefilms** [comment]
1) Close CommandPost
2) Open CommandPost
3) Open Preferences Panel
4) Go to the Touch Bar Panel
5) Click the first "Select" button
6) Right-click on the Chooser

This seems to create a new menubar item each time you right-click?

**Comment by latenitefilms** [comment]
It looks like it's the `self._rightClickMenubar` that's causing the issue for whatever reason.

Changing to `local rightClickMenubar` seems to fix things.

Thoughts?

**Comment by randomeizer** [comment]
Have you pushed your fix?

**Comment by randomeizer** [comment]
To a branch, I mean

**Comment by latenitefilms** [comment]
Changing it from `self._rightClickMenubar` to `local rightClickMenubar` isn't a proper fix.

However, I've also seen this error:

`2017-10-04 01:20:33: 01:20:33 ** Warning:   LuaSkin: hs.menubar:popupMenu() Missing menu object`

I wonder if this is the same problem as #787?

Maybe our choosers are just too complex?

**Comment by latenitefilms** [comment]
FYI: https://github.com/Hammerspoon/hammerspoon/issues/1567

**Comment by latenitefilms** [comment]
FYI - I can reproduce this with:

```
-- hs.chooser Example

local inspect = require("hs.inspect")

local logger = require("hs.logger")
logger.defaultLogLevel = 'debug'
log = logger.new("chooser")

local chooser = require("hs.chooser")
local menubar = require("hs.menubar")
local mouse = require("hs.mouse")

showAChoice = true
numberOfChoices = 100000

function choicesA()
	local result = {}

	for i=1, numberOfChoices, 1 do
		table.insert(result, {
			["text"] = "Choice " .. tostring(i),
			["subText"] = "This is the subtext of choice " .. tostring(i),
			["uuid"] = tostring(i)
		})
	end

	return result
end

function choicesB()
	local result = {}

	for i=1, numberOfChoices, 1 do
		table.insert(result, {
			["text"] = "New Choice " .. tostring(i),
			["subText"] = "This is the subtext of choice " .. tostring(i),
			["uuid"] = tostring(i)
		})
	end

	return result
end

myMenubar = menubar.new(false)
	:setMenu({
       {
       	title = "Refresh",
       	fn = function()
       		log.df("Refreshing Choices.")
       		if showAChoice then
       			choices = choicesB()
       		else
       			choices = choicesA()
       		end
       		showAChoice = not showAChoice
			myChooser:refreshChoicesCallback()
       	end
       },
   })

function completionFn(value)
	log.df("completionFn value: %s", inspect(value))
end

function choicesFn(value)
	if not choices then
		choices = choicesA()
	end
	return choices
end

function rightClickFn()
	local point = mouse.getAbsolutePosition()
	myMenubar:popupMenu(point)
end

function queryChangedFn(value)
	log.df("queryChangedFn value: %s", inspect(value))
end

function showFn()
	log.df("showFn triggered.")
end

myChooser = chooser.new(completionFn)
	:choices(choicesFn)
	--:queryChangedCallback(queryChangedFn)
	:showCallback(showFn)
	:rightClickCallback(rightClickFn)
	:show()
```

![screen shot 2017-10-05 at 5 11 45 pm](https://user-images.githubusercontent.com/22286696/31213099-5a843ab8-a9f0-11e7-8045-075ed2fe5ab2.png)

**Comment by latenitefilms** [comment]
Ummm... this seems to work fine though:

```
-- hs.chooser Example

local inspect = require("hs.inspect")

local logger = require("hs.logger")
logger.defaultLogLevel = 'debug'
log = logger.new("chooser")

local chooser = require("hs.chooser")
local menubar = require("hs.menubar")
local mouse = require("hs.mouse")

showAChoice = true
numberOfChoices = 100000

function choicesA()
	local result = {}

	for i=1, numberOfChoices, 1 do
		table.insert(result, {
			["text"] = "Choice " .. tostring(i),
			["subText"] = "This is the subtext of choice " .. tostring(i),
			["uuid"] = tostring(i)
		})
	end

	return result
end

function choicesB()
	local result = {}

	for i=1, numberOfChoices, 1 do
		table.insert(result, {
			["text"] = "New Choice " .. tostring(i),
			["subText"] = "This is the subtext of choice " .. tostring(i),
			["uuid"] = tostring(i)
		})
	end

	return result
end

myMenubar = menubar.new()
	:setMenu({
       {
       	title = "Refresh",
       	fn = function()
       		log.df("Refreshing Choices.")
       		if showAChoice then
       			choices = choicesB()
       		else
       			choices = choicesA()
       		end
       		showAChoice = not showAChoice
			myChooser:refreshChoicesCallback()
       	end
       },
   })
   :removeFromMenuBar()

function completionFn(value)
	log.df("completionFn value: %s", inspect(value))
end

function choicesFn(value)
	if not choices then
		choices = choicesA()
	end
	return choices
end

function rightClickFn()
	local point = mouse.getAbsolutePosition()
	myMenubar:popupMenu(point)
end

function queryChangedFn(value)
	log.df("queryChangedFn value: %s", inspect(value))
end

function showFn()
	log.df("showFn triggered.")
end

myChooser = chooser.new(completionFn)
	:choices(choicesFn)
	--:queryChangedCallback(queryChangedFn)
	:showCallback(showFn)
	:rightClickCallback(rightClickFn)
	:show()
```

**Comment by randomeizer** [comment]
So, what was the fix exactly?

**Comment by latenitefilms** [comment]
So the trick is you need to use` menubar.new()` (i.e. don't specify a boolean) and call `:removeFromMenuBar()` directly AFTER `:setMenu()`. It's clearly a bug - logged [here](https://github.com/Hammerspoon/hammerspoon/issues/1567).

**Comment by randomeizer** [comment]
Ah, ok. Did you check in the actual fix somewere? It's not in the log here.

**Comment by latenitefilms** [comment]
latenitefilms closed this in [0575983](https://github.com/CommandPost/CommandPost/commit/05759835b40cd98dcd0ab2984c06b98a504d8b63) 7 minutes ago
